nexus-writer/src/nexus/builder.rs

```diff
@@ -0,0 +1,237 @@
+        let args = RunStartArgs {
+            start_time,
+            run_name: Some(fbb.create_string(name)),
+            instrument_name: Some(fbb.create_string("Super MuSR")),
+            ..Default::default()
+        };
+        let message = RunStart::create(fbb, &args);
+        finish_run_start_buffer(fbb, message);
+        root_as_run_start(fbb.finished_data())
+    }
+
+    fn create_stop<'a, 'b: 'a>(
+        fbb: &'b mut FlatBufferBuilder,
+        name: &str,
+        stop_time: u64,
+    ) -> Result<RunStop<'a>, InvalidFlatbuffer> {
+        let args = RunStopArgs {
+            stop_time,
+            run_name: Some(fbb.create_string(name)),
+            ..Default::default()
+        };
+        let message = RunStop::create(fbb, &args);
+        finish_run_stop_buffer(fbb, message);
+        root_as_run_stop(fbb.finished_data())
+    }
+
+    #[test]
+    fn empty_run() {
+        let mut nexus = Nexus::new(None);
+        let mut fbb = FlatBufferBuilder::new();
+        let start = create_start(&mut fbb, "Test1", 16).unwrap();
+        nexus.start_command(start).unwrap();
+
+        assert_eq!(nexus.run_cache.len(), 1);
+        assert_eq!(nexus.run_cache[0].parameters().collect_from, 16);
+        assert!(nexus.run_cache[0]
+            .parameters()
+            .run_stop_parameters
+            .is_none());
+
+        fbb.reset();
+        let stop = create_stop(&mut fbb, "Test1", 17).unwrap();
+        nexus.stop_command(stop).unwrap();
+
+        assert_eq!(nexus.cache_iter().len(), 1);
+
+        let run = nexus.cache_iter().next();
+
+        assert!(run.is_some());
+        assert!(run.unwrap().parameters().run_stop_parameters.is_some());
+        assert_eq!(run.unwrap().get_name(), "Test1");
+
+        assert!(run.unwrap().parameters().run_stop_parameters.is_some());
+        assert_eq!(
+            run.unwrap()
+                .parameters()
+                .run_stop_parameters
+                .as_ref()
+                .unwrap()
+                .collect_until,
+            17
+        );
+    }
+
+    #[test]
+    fn no_run_start() {
+        let mut nexus = Nexus::new(None);
+        let mut fbb = FlatBufferBuilder::new();
+
+        let stop = create_stop(&mut fbb, "Test1", 0).unwrap();
+        assert!(nexus.stop_command(stop).is_err());
+    }
+
+    #[test]
+    fn no_run_stop() {
+        let mut nexus = Nexus::new(None);
+        let mut fbb = FlatBufferBuilder::new();
+
+        let start1 = create_start(&mut fbb, "Test1", 0).unwrap();
+        nexus.start_command(start1).unwrap();
+
+        fbb.reset();
+        let start2 = create_start(&mut fbb, "Test2", 0).unwrap();
+        assert!(nexus.start_command(start2).is_err());
+    }
+
+    #[test]
+    fn frame_messages_correct() {
+        let mut nexus = Nexus::new(None);
+        let mut fbb = FlatBufferBuilder::new();
+
+        let ts = GpsTime::new(0, 1, 0, 0, 16, 0, 0, 0);
+        let ts_start = {
```

Comment:
What is the purpose of the extra scope here?

Reply:
I think it was a relic from an earlier hack, that I don't need anymore. Removed.